game/tap.ksy

```diff
@@ -0,0 +1,49 @@
+meta:
+  id: tap
+  file-extension: tap
+  endian: le
+  license: CC0-1.0
+  title: ZX Spectrum tape file format
+doc-ref: https://faqwiki.zxnet.co.uk/wiki/TAP_format
+seq:
+  - id: block
+    type: block
+    repeat: eos
+enums:
+  flag_enum:
+    0x00: header
+    0xFF: data
+  header_type_enum:
+    0: program
+    1: num_array
+    2: char_array
+    3: code_file
+types:
+  block:
+    seq:
+      - id: length
+        contents: [0x13, 0x00]
+      - id: flag
+        type: u1
+        enum: flag_enum
+      - id: header
+        type: header_block
+      - id: data
+        size: header.data_length + 4
```

Comment:
If I understand this correctly, the `block` type currently parses a header block together with the following data block. This works fine if the data is complete, consistent and uses the standard structure, but would break in a few unusual cases:

* If the data length stored in the header block doesn't match the actual length of the following block, the `data` field will parse an incorrect amount of data, and any following blocks will not be parsed correctly.
* If the header block is for some reason followed by another header block, that header will be parsed as data instead of a header, which will also likely break parsing of the following blocks.
* If a data block is not preceded by a header, the data will be incorrectly parsed as a header instead (or more likely parsing will fail because the data length is not exactly 0x13).

The third case is probably the most important - I've seen some custom loaders use data blocks in the standard format, but without a preceding header block. These are valid tapes and should be accepted by the spec. The other two error cases are less important, because they only appear when the tape data is incomplete or out of order.

I think the best way to handle this would be to parse each TAP block on its own, without trying to link the header blocks with their corresponding data blocks. This should be fairly easy to do - each TAP block is preceded by its length (so there's no need to read the data block length from its header block) and includes a header/data type flag (so you can tell which blocks are headers and should be parsed).

Reply:
Great info. Would you have a sample TAP file demonstrating the third case?

Reply:
One example is the game [Full Throttle](http://www.worldofspectrum.org/infoseekid.cgi?id=0001908) - World of Spectrum has a [TAP download](http://www.worldofspectrum.org/pub/sinclair/games/f/FullThrottle.tap.zip) available. It has a custom loader (standard format, header and data) followed by a data block with no header.